File: Programmers/level2/게임_맵_최단거리/게임_맵_최단거리.py
from collections import deque


def solution(maps):
    x = [0, 0, -1, 1]
    y = [1, -1, 0, 0]
    col = len(maps)
    row = len(maps[0])
    vst = [[1] * row for _ in range(col)]
    vst[0][0] = 0

    b = deque([[0, 0, 1]])
    while b:  # bfs
        ny, nx, n = b.pop()
        if ny == col - 1 and nx == row - 1:
            return n
        for i in range(4):
            my = ny + y[i]
            mx = nx + x[i]
            if my < 0 or mx < 0 or my == col or mx == row:
                continue
            if maps[my][mx] == 1 and vst[my][mx] == 1:
                b.append([my, mx, n + 1])
                vst[my][mx] = 0
        #print(b)
    return -1

File: Programmers/level2/게임_맵_최단거리/게임_맵_최단거리.java
import java.util.*;

class Solution {
    public int solution(int[][] maps) {
        int answer = -1;
        int ly = maps.length;
        int lx = maps[0].length;
        int[][] d = {{0,1},{0,-1},{1,0},{-1,0}};

        Queue<int[]> q = new LinkedList<>();
        int[][] vst = new int[ly][lx]; // 0

        q.add(new int[]{0,0,1});
        vst[0][0] = 1;

        while (!q.isEmpty()) {
            int[] p = q.poll();

            if (p[0] == ly-1 && p[1] == lx-1) {
                return p[2];
            }
            //System.out.print(p[0] + "" +  p[1]);

            for (int dr[] : d) {
                int y = p[0] + dr[0];
                int x = p[1] + dr[1];

                if (0 <= y && y < ly && 0 <= x && x < lx && maps[y][x] == 1 && vst[y][x] == 0) {
                    vst[y][x] = 1;
                    q.add(new int[]{y, x, p[2]+1});
                }
            }
        }

        return answer;
    }
}